src/node/txreconciliation.cpp

```diff
@@ -75,7 +81,37 @@ class TxReconciliationState
      */
     std::map<uint32_t, Wtxid> m_short_id_mapping;
 
-    TxReconciliationState(bool we_initiate, uint64_t k0, uint64_t k1) : m_we_initiate(we_initiate), m_k0(k0), m_k1(k1) {}
+    TxReconciliationState(bool we_initiate, uint64_t k0, uint64_t k1) : m_we_initiate(we_initiate), m_k0(k0), m_k1(k1), m_delayed_local_set(), m_local_set(0, m_delayed_local_set.hash_function()) {}
+
+    /**
+    * Checks whether a transaction is already in the set. If `include_delayed` is set, the delayed set is also
+    * checked. Otherwise, transactions are only looked up in the regular set.
+    */
+    bool ContainsTx(const Wtxid& wtxid, bool include_delayed)
+    {
+        bool found = m_local_set.find(wtxid) != m_local_set.end();
+        if (include_delayed) {
+            found |= m_delayed_local_set.find(wtxid) != m_delayed_local_set.end();
+        }
+
+        return found;
+    }
+
+    /**
+     * Computes the size of the reconciliation set (including both available and delayed transactions)
+     */
+    size_t ReconSetSize()
+    {
+        return m_local_set.size() + m_delayed_local_set.size();
+    }
+
+    bool RemoveFromSet(const Wtxid& wtxid)
+    {
+        auto r = m_local_set.erase(wtxid) + m_delayed_local_set.erase(wtxid);
```

Comment:
Nit

```
        const size_t removed = m_local_set.erase(wtxid) + m_delayed_local_set.erase(wtxid);
        // Data must be in one of the sets at most
        Assume(removed <= 1);
        return removed == 1;
```

Reply:
Not too sure about the `return removed == 1`

The assume is there to make sure that if the code happens to be wrong, it is catched on testing. In the current approach, if the code moving data from `delayed_set` to `m_local_set` was broken somehow, this method will return that data was actually removed. If were to return `removed == 1` in the aforementioned case, it would return false, and the caller's logic may break.

I don't think any of these cases is likely to happen given the assume will catch it before getting merged, but the proposed change makes me anxious.

PS: I agree on doing the variable renaming if you feel that makes the code more readable though

Reply:
I thought `size_t 0` will be cast to `bool false` on return, and just suggested the exact same behavior (1==true, 0==false) but without forcing the reader to think about casting and cause this exact confusion between us.
Am i wrong?

-------

Ahhh, so you saying `2==true` is also important? Then I'll be fine with explicit `return removed >= 1`, and comment that `2` should not normally happen.

Reply:
`2==true` would mean that we had the transaction in both places, which should never happen (I'd mean that a transaction is both ready and delayed at the same time).

If that were to happen in production for whatever reason, this would return false with the suggested changes, which may break something?

Reply:
Probably a taste thing. I think `auto` should be used only when it's obvious what type it is.

Here there are two non-obvious occasions:
- `auto x = bool + bool` resulting in x=2.
- casting 2 to bool for the result.

I won't insist but i think verbose types are better here. Feel free to close.

Reply:
That is fair, I'll change it to bool to be explicit